File: numerov/eigenvalues.py
# Algorithm to find numeric values for the energy eigenvalues
# of a given potential function and particle.

if __name__ == 'numerov.eigenvalues':
  from . import numerov
  from . import root
  from . import units
else:
  import numerov
  import root
  import units
import numpy as np
import math

def energy(f, m, x_start, x_end, y_start, y_end, step_count=1000, max_iterations=1000):
  """Find numeric solutions to the energy eigenvalues
     of a given system.
  
  @param f              callable (potential function in Joules)
  @param m              float    (mass of particle in kg)
  @param x_start        float    (start of integration range)
  @param x_end          float
  @param y_start        float    (boundary condition on x_start)
  @param y_end          float
  @param step_count     int      (steps in numerov integration)
  @param max_iterations int      (max number of energies to return)

  @return [float] (energy eigenvalues in scaled units)
  """
  # Find constant values
  half_length = units.scaleL(x_end-x_start, m) / 2
  energies = []
  norm = 1.0
  def f_n(E, back=False):
    if back:
      return lambda x: units.scaleE(f(x_end - units.unscaleL(x, m)), m) - E
    return lambda x: units.scaleE(f(x_start + units.unscaleL(x, m)), m) - E
  def f_r(E):
    nonlocal norm
    # Perform Numerov integration
    y_vals_f = numerov.integrate(y_start, norm, f_n(E, False), 0, half_length, step_count=step_count//2)
    y_vals_b = numerov.integrate(y_start, norm, f_n(E, True), 0, half_length, step_count=step_count//2)
    # Make continuous
    cont_factor = y_vals_f[len(y_vals_f)-1] / y_vals_b[len(y_vals_b)-1]
    # Compute derivatives
    d_f = (y_vals_f[len(y_vals_f)-1] - y_vals_f[len(y_vals_f)-2]) * step_count / 2 / half_length
    d_b = (y_vals_b[len(y_vals_b)-2] - y_vals_b[len(y_vals_b)-1]) * cont_factor * step_count / 2 / half_length
    # Improve normalization
    norm = 10 * norm / np.max(np.abs(y_vals_f))
    return d_f - d_b
  # Find energies using the root / shooting method
  bounds = (0,0)
  i = 0
  while bounds != False and i < max_iterations:
    print("Iteration " + str(i))
    bounds = (bounds[1], bounds[1]+1e-10)
    # Note: Think about better / more intelligent step sizes vs how many steps needed
    bounds = root.expandBrackets(f_r, bounds[0], bounds[1], 1.+1e-3, 1e-1, 1<<15)
    if bounds == False:
      break
    energy = root.bracket(f_r, bounds[0], bounds[1])
    if energy != False:
      energies.append(energy)
    elif i == 0:
      # If this is the first, add one to match c implementation
      max_iterations += 1
    i += 1
  return energies

def psi(f, m, E, x_start, x_end, y_start, y_end, step_count=1000):
  """Find the plot for Psi for a given energy
  
  @param f              callable (potential function in Joules)
  @param m              float    (mass of particle in kg)
  @param E              float    (scaled energy eigenvalue to plot)
  @param x_start        float    (start of integration range)
  @param x_norm         float    (normalization parameter, see energy)
  @param y_start        float    (boundary condition on x_start)
  @param y_end          float
  @param step_count     int      (steps in numerov integration)

  @return list of float (values of psi, not normalized)
  """
  # Find constant values
  half_length = units.scaleL(x_end-x_start, m) / 2
  energies = []
  def f_n(E, back=False):
    if back:
      return lambda x: units.scaleE(f(x_end - units.unscaleL(x, m)), m) - E
    return lambda x: units.scaleE(f(x_start + units.unscaleL(x, m)), m) - E
  # Perform Numerov integration
  y_vals_f = numerov.integrate(y_start, 1.0, f_n(E, False), 0, half_length, step_count=step_count//2)
  y_vals_b = numerov.integrate(y_end, 1.0, f_n(E, True), 0, half_length, step_count=step_count//2)
  # Make continuous
  y_vals_b *= y_vals_f[len(y_vals_f)-1] / y_vals_b[len(y_vals_b)-1]
  # Return normalized result
  return np.concatenate((y_vals_f, np.flip(y_vals_b, 0))) / np.max(np.abs(y_vals_f))

File: numerov_c/eigenvalues.c
#import "eigenvalues.h"

#define INT_SHORTHAND(E) integrate(E, f, &m, &x_start, &x_end, &y_start, &y_end, &step_count, &step_width, &norm)

#define BRACKET_THRESHOLD 1e-3l

#define ROOT_GEOM (1l+1e-3l)
#define ROOT_STEP 1e-1
#define ROOT_ACCURACY 1e-10l
#define ROOT_MAX_ITERATIONS 1<<10


double callPythonPotential(PyObject *f, double x) {
  // Get value from python defined potential
  PyObject *potential_py = PyObject_CallFunction(f, "d", x);
  if (potential_py == NULL) {
    // Raise an error
    PyErr_BadArgument();
    return 0l;
  }
  // Return the float value (but don't do error checking)
  return PyFloat_AsDouble(potential_py);
}

// Integration function needed by the numerov
// Why this horrid function that gets passed the local variables? Because I need the procedure twice.
// NOTE: This is called using the above macro INT_SHORTHAND
double integrate(double E, PyObject *f, double *m, double *x_start, double *x_end, double *y_start, double *y_end, int *step_count, double *step_width, double *norm) {
  // k / kk def as in step fn
  double psi_k_f, psi_kk_f, psi_k_b, psi_kk_b, psi_temp, psi_max, x;
  // Initialize variables
  x = *step_width;
  psi_k_f = *norm;
  psi_k_b = *norm;
  psi_kk_f = *y_start;
  psi_kk_b = *y_end;
  psi_max = 0l;
  // Step forward / backward through integration range
  for (int k = 0; k < *step_count / 2; k ++, x += *step_width) {
    // Forward step
    psi_temp = psi_k_f;
    psi_k_f = numerov_step(
      psi_k_f,
      psi_kk_f,
      units_scaleE(callPythonPotential(f, *x_start + units_unscaleL(x - *step_width, *m)), *m) - E,
      units_scaleE(callPythonPotential(f, *x_start + units_unscaleL(x, *m)), *m) - E,
      units_scaleE(callPythonPotential(f, *x_start + units_unscaleL(x + *step_width, *m)), *m) - E,
      *step_width
    );
    psi_kk_f = psi_temp;
    // Backward step
    psi_temp = psi_k_b;
    psi_k_b = numerov_step(
      psi_k_b,
      psi_kk_b,
      units_scaleE(callPythonPotential(f, *x_end - units_unscaleL(x - *step_width, *m)), *m) - E,
      units_scaleE(callPythonPotential(f, *x_end - units_unscaleL(x, *m)), *m) - E,
      units_scaleE(callPythonPotential(f, *x_end - units_unscaleL(x + *step_width, *m)), *m) - E,
      *step_width
    );
    psi_kk_b = psi_temp;
    // Keep max for normalization (just from first one for convenience, since backwards is scaled after)
    if (fabs(psi_k_f) > psi_max) {
      psi_max = fabs(psi_k_f);
    }
  }
  // Adjust normalization
  *norm = 10l * *norm / psi_max;
  // Return difference in derivatives (and match backwards to be continuous)
  return (psi_k_f - psi_kk_f + (psi_k_b - psi_kk_b) * psi_k_f / psi_k_b) / *step_width;
}

// Find energy eigenvalues (returns python list)
PyObject *eigenvalues_energy(
  PyObject *f,       // Potential function in Joules (takes meters)
  double m,          // Mass of particle
  double x_start,    // Start of integration range
  double x_end,
  double y_start,    // Boundary condition at x_start
  double y_end,
  int step_count,    // Number of integration steps
  int max_iterations // Max number of energies to return
) {
  // Test if f is a function
  if (!PyCallable_Check(f)) {
    PyErr_BadArgument();
    return NULL;
  }
  if (max_iterations <= 0) {
    max_iterations = 1<<2;
  }
  // Determine constant values
  double step_width = units_scaleL(x_end-x_start, m) / (double)step_count;

  // Variables for the Numerov function
  double norm = 1l;

  // Variables for the bounds and energies, reference need not be decreased, since
  // we pass on the object
  PyObject* energies_list = PyList_New(0);
  if (energies_list == NULL) {
    return PyErr_NoMemory();
  }
  double lower, upper = 0l, energy_res;
  double v_upper, v_lower, midpoint;
  int root_success;

  // Variables for the integration, bracketing and root finding
  // NOTE: This won't abort if no bounds are found, but continue to
  //       the next bracketing step.
  for (int i_energy = 0; i_energy < max_iterations; i_energy ++) {
    // Set new bounds
    lower = upper;
    upper += BRACKET_THRESHOLD;
    // Expand energy brackets
    v_lower = INT_SHORTHAND(lower);
    root_success = 0;
    for (int i = 0; i < ROOT_MAX_ITERATIONS; i ++) {
      v_upper = INT_SHORTHAND(upper);
      if (v_upper * v_lower <= 0) {
        root_success = 1;
        break;
      } else {
        upper += pow(ROOT_GEOM, (double)i) * ROOT_STEP;
      }
    }
    if (!root_success) {
      continue;
    }
    // Bracket energies to find roots
    root_success = 0;
    for (int i = 0; i < ROOT_MAX_ITERATIONS; i ++) {
      v_upper = INT_SHORTHAND(upper);
      midpoint = (upper+lower) / 2l;
      // Adjust bounds
      if (INT_SHORTHAND(midpoint) * v_upper <= 0l) {
        // The zero is between MID <-> UPPER
        lower = midpoint;
      } else {
        // The zero is between LOWER <-> MID
        upper = midpoint;
      }
      // Test if we reached the desired accuracy
      if (upper - lower < ROOT_ACCURACY) {
        root_success = 1;
        energy_res = (upper+lower) / 2l;
      }
    }
    if (!root_success || energy_res < BRACKET_THRESHOLD) {
      // Discard if no solution is found / energy is below threshold (ie E=0)
      if (root_success && energy_res < BRACKET_THRESHOLD) {
        // Do an extra one to better reflect argument description
        max_iterations += 1;
      }
      continue;
    }
    // Store energy
    // NOTE: Reference need not be decreased, since we pass on the values
    PyObject* energy_py = Py_BuildValue("d", energy_res);
    if (energy_py != NULL) {
      PyList_Append(energies_list, energy_py);
    } else {
      // Raise error with python
      PyErr_NoMemory();
    }
  }
  // We're done
  return energies_list;
}